docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -0,0 +1,502 @@
+  * [Open Questions](#open-questions)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+    + [1. CMPs are under-utilized](#1-cmps-are-under-utilized)
+    + [2. Decisions about config management tools are limited by the core code](#2-decisions-about-config-management-tools-are-limited-by-the-core-code)
+    + [3. Ksonnet is deprecated, and CMPs are a good place to maintain support](#3-ksonnet-is-deprecated-and-cmps-are-a-good-place-to-maintain-support)
+    + [Goals](#goals)
+    + [Non-Goals](#non-goals)
+  * [Proposal](#proposal)
+    + [Use cases](#use-cases)
+      - [Use case 1: building Argo CD without config management dependencies](#use-case-1-building-argo-cd-without-config-management-dependencies)
+      - [Use case 2: writing CMPs with rich UI experiences](#use-case-2-writing-cmps-with-rich-ui-experiences)
+    + [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      - [Prerequisites](#prerequisites)
+      - [Terms](#terms)
+      - [CMP config schema](#cmp-config-schema)
+      - [Parameters manifest / parameters serialization format](#parameters-manifest--parameters-serialization-format)
+      - [Parameter definition schema](#parameter-definition-schema)
+      - [Parameters announcement schema](#parameters-announcement-schema)
+      - [Parameter list schema](#parameter-list-schema)
+    + [Detailed examples](#detailed-examples)
+      - [Example 1: trivial parameterized CMP](#example-1-trivial-parameterized-cmp)
+      - [Example 2: Helm parameters from Kustomize dependency](#example-2-helm-parameters-from-kustomize-dependency)
+      - [Example 3: simple Helm CMP](#example-3-simple-helm-cmp)
+    + [Security Considerations](#security-considerations)
+      - [Increased scripting](#increased-scripting)
+    + [Risks and Mitigations](#risks-and-mitigations)
+    + [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  * [Drawbacks](#drawbacks)
+  * [Alternatives](#alternatives)
+
+## Open Questions
+
+Is there a way to make it easier for the generate command to consume parameters? (Easier than parsing a JSON object, 
+that is.)
+* We could translate all parameters to env vars, like `ARGOCD_PARAM_{section}_{param name}`.
+  * Advantage: no need to pull the param from JSON.
+  * Disadvantage: we'd need an easy-to-predict and collision-resistant way to translate arbitrary parameter names
+    to env vars. 
+    * Option 1: Enforce some parameter name convention which allows a 1:1 translation to env
+      vars. Is limiting the param namespace worth it for the usability win?
+    * Option 2: Only translate to an env var if the param name happens to be in the safely-translatable
+      namespace. For example, `set-file` becomes `SET_FILE`, but `omg!!!param` doesn't get translated to an env var.
+    * Option 3: Add an `envVar` field to the [parameter definition schema](#parameter-definition-schema) to specify
```

Comment:
If the plugin doesn't control the parameter names, it might have a hard time specifying the `envVar` field.

Reply:
@jopit agreed. This would basically kick the can down the road to the plugin author. I guess one could argue that the plugin author might have more knowledge of the parameter names space and be able to write a better name -> envVar algorithm than we could.